### PR TITLE
Fix issues with document state getting out of sync when changing and saving

### PR DESCRIPTION
**Pull request by rchl**
debounced call was incorrectly running on the sync thread causing potential
race issues when "pending_changes" was modified/checked from different
threads.

Also, made text document version reported in did_change notification
represents document version at the time of the change rather than at the
time of sending the notification. Without it, we sometimes have sent
consecutive notifications with different contents but the same text document
version. Which I imagine servers don't like.

### CODE REVIEW

**Comment by rchl** [comment]
I'd like to try to make a test for this but I don't have time for it now so please hold off with integrating for now.

**Comment by rchl** [comment]
9e56b3b73de22afc48a8d71df212890dd25cff09 can be reproduced with something like that:

```diff
--- a/plugin/documents.py
+++ b/plugin/documents.py
@@ -652,6 +657,7 @@ class DocumentSyncListener(sublime_plugin.ViewEventListener, AbstractViewListene
             debug("couldn't find a text change listener for", self)
             return
         self._text_change_listener_found = True
+        sublime.set_timeout(lambda: self.view.run_command('insert', {'characters': '123'}))
         self.manager.register_listener_async(self)
         views = buf.views()
         if not isinstance(views, list):
```
and re-opening the file

**Comment by predragnikolic** [comment]
I am really happy to say that these changes fixes the error that I got when modifying the view while the server was initializing :)

**Comment by rwols** [comment]
Is this done?

**Comment by rchl** [comment]
Probably better to get this in than waiting for me to try to figure out how to test it as it would be non-trivial and I didn't manage to do it in 1h for the second issue at least.

I've just cleaned the history and would suggest merging those without squashing as those are two separate issues.

File: plugin/documents.py [hunk 1]
```diff
@@ -645,7 +651,6 @@ def _register_async(self) -> None:
         if not text_change_listener:
             debug("couldn't find a text change listener for", self)
             return
-        text_change_listener.view_listeners.add(self)
         self._text_change_listener_found = True
```
Comment (rwols) [inline plugin/documents.py:654]: This variable name, _text_change_listener_found, is now kind of confusing. Can we rename it to `_registered` or something?

Comment (rwols) [reply plugin/documents.py:654]: All it really needs to do is to prevent registering the view to the window over and over again.

Comment (rchl) [reply plugin/documents.py]: Changed to `_registered`.

**Review by rwols — APPROVED** [review]

**Pull request merged**